typing/typecore.ml

```diff
@@ -4902,10 +4902,25 @@ and type_let ?(check = fun s -> Warnings.Unused_var s)
     end else pat_list in
   (* Only bind pattern variables after generalizing *)
   List.iter (fun f -> f()) force;
+  let sexp_is_fun { pvb_expr = sexp; _ } =
+    match sexp.pexp_desc with
+    | Pexp_fun _ | Pexp_function _ -> true
+    | _ -> false
+  in
   let exp_env =
     if is_recursive then new_env
-    else if not is_recursive then begin
-      (* add ghost bindings to help detecting missing "rec" keywords *)
+    else if not is_recursive && List.for_all sexp_is_fun spat_sexp_list
```

Comment:
Do we want `List.for_all` or `List.exists`? I would be content with a `let .. and ..`, one of them being a function.

Reply:
I do not have a strong opinion either way. I guess it should be relatively safe to use `List.exists` since `let .. and ..` is rarely used to define something else than recursive functions.